crates/synthesis/src/multi_controlled/mcx.rs

```diff
@@ -414,3 +441,510 @@ pub fn synth_mcx_noaux_v24(py: Python, num_controls: usize) -> PyResult<CircuitD
+
+    // Handle small cases explicitly
+    if n == 2 {
+        circuit.cx(0, 1);
+    } else {
+        circuit.h(num_controls as u32);
+
+        // The construction described in Fig.7 of the paper only works for even values of n.
+        // The construction described in Fig.8 of the paper works for all values of n and is better than the one
+        // in Fig.7 when n<23.
+
+        if (n % 2 == 0) && (n >= 23) {
+            // This implements C^{n-1}(V) in Fig.7.
+
+            // These implement U^{n-1}_{+1} and U^{n-1}_{-1} (last qubit is ancilla)
+            let increment_plus_1 = increment_1_dirty((n - 1) as u32, true)?;
+            let increment_minus_1 = increment_1_dirty((n - 1) as u32, false)?;
+            let increment_qubits: Vec<Qubit> = (0..n).map(|q| Qubit(q as u32)).collect();
+
+            circuit.compose(&increment_plus_1, &increment_qubits, &[])?;
+
+            // This implements U^{n-1}_{V_{n-1}}(-1)
+            let mut phi = -PI;
+            for q in (1..n - 1).rev() {
+                phi /= 2.;
+                circuit.cp(phi, q as u32, (n - 1) as u32);
+            }
+
+            circuit.compose(&increment_minus_1, &increment_qubits, &[])?;
+
+            // This implements U^{n-1}_{V_{n-1}}(-1)
+            let mut phi = PI;
+            for q in (1..n - 1).rev() {
+                phi /= 2.;
+                circuit.cp(phi, q as u32, (n - 1) as u32);
+            }
+
+            circuit.cp(phi, 0, (n - 1) as u32);
+        } else {
+            // This implements C^{n-1}(V) in Fig.8.
+
+            // These implement U^{n-1}_{+1} and U^{n-1}_{-1} (last qubit is ancilla)
+            let increment_plus_1 = increment_2_dirty((n - 2) as u32, true)?;
+            let increment_minus_1 = increment_2_dirty((n - 2) as u32, false)?;
+            let increment_qubits: Vec<Qubit> = (0..n).map(|q| Qubit(q as u32)).collect();
+
+            circuit.compose(&increment_plus_1, &increment_qubits, &[])?;
+
+            // This implements C(U^{n-2}_{V_{n-2}}(-1)) in Fig.8.
+            let mut phi = -PI;
+            for q in (1..n - 2).rev() {
+                phi /= 2.;
+                circuit.ccp(phi, q as u32, (n - 2) as u32, (n - 1) as u32);
+            }
+
+            circuit.compose(&increment_minus_1, &increment_qubits, &[])?;
+
+            // This implements C(U^{n-2}_{V_{n-2}}(1)) in Fig.8.
+            let mut phi = PI;
+            for q in (1..n - 2).rev() {
+                phi /= 2.;
+                circuit.ccp(phi, q as u32, (n - 2) as u32, (n - 1) as u32);
+            }
+
+            circuit.ccp(phi, 0, (n - 2) as u32, (n - 1) as u32);
+        }
+
+        circuit.h(num_controls as u32);
+    }
+
+    Ok(circuit)
+}
+
+#[cfg(test)]
+mod test {
+    use approx::abs_diff_eq;
+    use qiskit_quantum_info::unitary_sim::sim_unitary_circuit;
+
+    use super::{increment_n_dirty_large, increment_n_dirty_small};
+
+    #[test]
+    fn test_increment_n_dirty() {
```

Comment:
it's great to see these tests! 
- perhaps the range of `nq` could be larger? we usually take 1,2,...,8 (in most tests)
- are there tests needed also for `increment_1_dirty` and `increment_2_dirty` ?
- are there tests needed for `synth_relative_mcx` ?